README.md
# nbid

[![codecov](https://codecov.io/gh/szkiba/nbid/branch/master/graph/badge.svg)](https://codecov.io/gh/szkiba/nbid)
[![GoDoc](https://img.shields.io/badge/pkg.go.dev-doc-blue)](http://pkg.go.dev/github.com/szkiba/nbid)
[![Go Report Card](https://goreportcard.com/badge/github.com/szkiba/nbid)](https://goreportcard.com/report/github.com/szkiba/nbid)

**Name Based globally unique ID generator**

**NBID** can be generate from arbitrary names using cryptographic hash function (SHA256 by default). The binary representation of the NBID is a 16 byte byte array. The string representation is using base32 hex (w/o padding) for space efficiency when stored in that form (26 bytes). The hex variant of base32 is used to retain the sortability (string and binary representation has same order).

NBID doesn't use base64 because case sensitivity and the 2 non alphanum chars may be an issue when transported as a string between various systems. To validate a base32 NBID, expect a 26 chars long, all uppercase sequence of `A` to `V` letters and `0` to `9` numbers (`[0-9A-V]{26}`).

## Features

- *Easy to use* - Simly a 128 bit hash without complexity of UUID.
- *Modern hash* - Default hash function is SHA256, but you can use any hash with at least 128 bit output.
- *Compact* - String representation size is only 26 characters.
- *Sortable* - Binary and string representation has same order (because of base32hex).
- *No case sensitivity* - Using only uppercase letters and numbers in string representaiton make NBID usable in environments without case sensitive names (DNS host names, FAT filesystem, etc).

## Install

> This section about CLI tool, for API usage check [Documentation](https://pkg.go.dev/github.com/szkiba/nbid).

You can install the pre-compiled binary or use Docker.

### Install the pre-compiled binary

Download the pre-compiled binaries from the [releases page](https://github.com/szkiba/nbid/releases) and
copy to the desired location.

### Install with Go

If you have Go environment set up, you can build nbid from source by running:

```sh
go get github.com/szkiba/nbid/cmd/nbid
```

Binary would be installed to $GOPATH/bin/nbid.

### Running with Docker

You can also use it within a Docker container. To do that, you'll need to
execute the following:

```sh
docker run szkiba/nbid
```

### Verifying your installation

To verify your installation, use the `nbid -v` command:

```sh
$ nbid -v

nbid/1.0.0 linux/amd64
```

You should see `nbid/VERSION` in the output.

### Usage

To print usage information, use the `nbid --help` command:

```
$ nbid --help

usage: nbid [name]

Generate NBID for name, or random NBID if name is missing.

Example: nbid "The quick brown fox jumps over the lazy dog"
Output: QUKFNCO7QU098QEAJAUB021E9S

  -v    prints version
```

## TODO

Document, document, document...


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<a name="unreleased"></a>
## [Unreleased]


<a name="v1.0.1"></a>
## [v1.0.1] - 2021-03-16
### Docs
- fix some little documentation mistakes


<a name="v1.0.0"></a>
## v1.0.0 - 2021-03-14
### Chore
- prepare v1.0.0

### Feat
- Initial version


[Unreleased]: https://github.com/szkiba/yare/compare/v1.0.1...HEAD
[v1.0.1]: https://github.com/szkiba/yare/compare/v1.0.0...v1.0.1


id_test.go
// MIT License
//
// Copyright (c) 2021 Iván Szkiba
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package nbid_test

import (
	"crypto/sha256"
	"encoding/hex"
	"testing"

	"github.com/szkiba/nbid"

	"github.com/stretchr/testify/assert"
)

func TestMustParse(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name      string
		in        string
		out       string
		wantPanic bool
	}{
		{name: "normal", in: "ABCDEFGHIJKLMNOPQRSTUV1234", out: "ABCDEFGHIJKLMNOPQRSTUV1234"},
		{name: "error_size", in: "small", wantPanic: true},
		{name: "error_value", in: "XXXXXXXXXXXXXXXXXXXXXXXXXX", wantPanic: true},
	}
	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			if tt.wantPanic {
				defer func() {
					if r := recover(); r == nil {
						t.Errorf("The code did not panic")
					}
				}()
			}

			id := nbid.MustParse(tt.in)

			txt, err := id.MarshalText()
			assert.Nil(t, err)
			assert.Equal(t, tt.out, string(txt))
		})
	}
}

func TestNew(t *testing.T) {
	t.Parallel()

	data := []byte("The quick brown fox jumps over the lazy dog")

	id := nbid.New(data)
	idHash := nbid.NewHash(sha256.New(), data)

	assert.NotEqual(t, nbid.Nil, id)
	assert.Equal(t, idHash, id)

	assert.Equal(t, "QUKFNCO7QU098QEAJAUB021E9S", id.String())
}

func TestRandom(t *testing.T) {
	t.Parallel()

	id1 := nbid.Random()
	id2 := nbid.Random()

	assert.NotEqual(t, nbid.Nil, id1)
	assert.NotEqual(t, nbid.Nil, id2)
	assert.NotEqual(t, id1, id2)
}

func TestIsNil(t *testing.T) {
	t.Parallel()

	var id nbid.NBID

	assert.True(t, id.IsNil())
	assert.False(t, nbid.Random().IsNil())
}

func TestCompareAndEqual(t *testing.T) {
	t.Parallel()

	a := nbid.NBID{1}
	b := nbid.NBID{2}
	c := nbid.NBID{1}

	assert.True(t, a.Equal(c))
	assert.False(t, a.Equal(b))

	assert.Less(t, a.Compare(b), 0)
	assert.Greater(t, b.Compare(a), 0)
	assert.Equal(t, a.Compare(c), 0)
}

func TestText(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name    string
		in      string
		out     string
		wantErr bool
	}{
		{name: "normal", in: "ABCDEFGHIJKLMNOPQRSTUV1234", out: "ABCDEFGHIJKLMNOPQRSTUV1234"},
		{name: "zero", in: "AAAAAAAAAAAAAAAAAAAAAAAAA8", out: "AAAAAAAAAAAAAAAAAAAAAAAAA8"},
		{name: "error_size", in: "small", wantErr: true},
		{name: "error_value", in: "XXXXXXXXXXXXXXXXXXXXXXXXXX", wantErr: true},
		{name: "sample1", in: "DEJRG44TLK8T305K0304VL1GP0", out: "DEJRG44TLK8T305K0304VL1GP0"},
		{name: "sample2", in: "DEJRG4CTLK8T305K0304VL1GP0", out: "DEJRG4CTLK8T305K0304VL1GP0"},
	}
	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()
			id, err := nbid.Parse(tt.in)

			if tt.wantErr {
				assert.Error(t, err)

				return
			}

			txt, err := id.MarshalText()
			assert.Nil(t, err)
			assert.Equal(t, tt.out, string(txt))
		})
	}
}

func TestBinary(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name    string
		hex     string
		wantErr bool
	}{
		{name: "normal", hex: "e12340c084a111ebb6a7af57dd3dfeda"},
		{name: "zero", hex: "00000000000000000000000000000000"},
		{name: "error_size", hex: "42", wantErr: true},
	}
	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			b, err := hex.DecodeString(tt.hex)

			assert.Nil(t, err)

			id1, err1 := nbid.FromBytes(b)

			var id2 nbid.NBID
			err2 := id2.UnmarshalBinary(b)

			if tt.wantErr {
				assert.Error(t, err1)
				assert.Error(t, err2)

				return
			}

			// FromBytes and UnmarshalBinary should result same id
			assert.Equal(t, b, id1[:])
			assert.Equal(t, b, id2[:])

			assert.Equal(t, b, id1.Bytes())
			assert.Equal(t, b, id2.Bytes())

			// MarshalBinary should result same byte array
			b1, err := id1.MarshalBinary()
			assert.Nil(t, err)

			b2 := id1.Bytes()

			assert.Equal(t, b, b1)
			assert.Equal(t, b, b2)

			assert.True(t, id1.Equal(id2))
			assert.True(t, id2.Equal(id1))
		})
	}
}


sql_test.go
// MIT License
//
// Copyright (c) 2021 Iván Szkiba
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package nbid_test

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/szkiba/nbid"
)

func TestScan(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name    string
		src     interface{}
		want    nbid.NBID
		wantErr bool
	}{
		{
			name: "normal",
			src:  "QUKFNCO7QU098QEAJAUB021E9S",
			want: nbid.MustParse("QUKFNCO7QU098QEAJAUB021E9S"),
		},
		{name: "nil", src: nil, want: nbid.Nil},
		{name: "empty_string", src: "", want: nbid.Nil},
		{
			name:    "invalid_string",
			src:     "XXX",
			wantErr: true,
		},
		{name: "empty_bytes", src: []byte{}, want: nbid.Nil},
		{
			name: "bytes_string",
			src:  []byte("QUKFNCO7QU098QEAJAUB021E9S"),
			want: nbid.MustParse("QUKFNCO7QU098QEAJAUB021E9S"),
		},
		{
			name: "bytes",
			src:  nbid.MustParse("QUKFNCO7QU098QEAJAUB021E9S").Bytes(),
			want: nbid.MustParse("QUKFNCO7QU098QEAJAUB021E9S"),
		},
		{name: "invalid_type", src: 42, wantErr: true},
	}
	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()

			var id nbid.NBID

			err := id.Scan(tt.src)

			if tt.wantErr {
				assert.Error(t, err)

				return
			}

			assert.Nil(t, err)
			assert.Equal(t, tt.want, id)
		})
	}
}

func TestValue(t *testing.T) {
	t.Parallel()

	id := nbid.Random()

	val, err := id.Value()
	assert.Nil(t, err)
	assert.Equal(t, id.String(), val)
}


json_test.go
// MIT License
//
// Copyright (c) 2021 Iván Szkiba
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package nbid_test

import (
	"encoding/json"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/szkiba/nbid"
)

func TestJSON(t *testing.T) {
	t.Parallel()

	type x struct {
		ID nbid.NBID `json:"id"`
	}

	tests := []struct {
		name string
		json string
		id   string
	}{
		{
			name: "normal",
			json: `{"id":"ABCDEFGHIJKLMNOPQRSTUV1234"}`,
			id:   "ABCDEFGHIJKLMNOPQRSTUV1234",
		},
		{
			name: "null",
			json: `{"id":null}`,
		},
	}
	for _, tt := range tests {
		tt := tt
		t.Run(tt.name, func(t *testing.T) {
			t.Parallel()
			var id nbid.NBID

			if tt.id != "" {
				assert.Nil(t, id.UnmarshalText([]byte(tt.id)))
			}

			data := x{}
			err := json.Unmarshal([]byte(tt.json), &data)

			assert.Nil(t, err)
			assert.Equal(t, id, data.ID)

			b, err := json.Marshal(data)
			assert.Nil(t, err)

			assert.Equal(t, tt.json, string(b))
		})
	}
}


doc.go
// MIT License
//
// Copyright (c) 2021 Iván Szkiba
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

// Package nbid is a Name Based globally unique ID generator.
//
// NBID is generated from names using cryptographic hash function (SHA256 by default).
// The binary representation of the NBID is a 16 byte byte array.
// The string representation is using base32 hex (w/o padding) for space efficiency
// when stored in that form (26 bytes). The hex variant of base32 is used to retain the
// sortability (string and binary representation has same order).
//
// NBID doesn't use base64 because case sensitivity and the 2 non alphanum chars may be an
// issue when transported as a string between various systems. To validate a base32 NBID,
// expect a 26 chars long, all uppercase sequence of `A` to `V` letters and `0` to `9` numbers
// (`[0-9A-V]{26}`).
//
package nbid


sql.go
// MIT License
//
// Copyright (c) 2021 Iván Szkiba
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package nbid

import (
	"database/sql/driver"
	"fmt"
	"reflect"
)

var invalidValue = reflect.Value{}

// Scan implements sql.Scanner so NBIDs can be read from databases transparently.
// Currently, database types that map to string and []byte are supported.
func (id *NBID) Scan(src interface{}) error {
	switch src := src.(type) {
	case nil:
		return nil

	case string:
		if src == "" {
			return nil
		}

		u, err := Parse(src)
		if err != nil {
			return err
		}

		*id = u

	case []byte:
		if len(src) == 0 {
			return nil
		}

		if len(src) != rawLen {
			return id.Scan(string(src))
		}

		copy((*id)[:], src)

	default:
		return fmt.Errorf("%w: unable to scan type %T", ErrInvalidID, src)
	}

	return nil
}

// Value implements sql.Valuer so that NBIDs can be written to databases transparently.
func (id NBID) Value() (driver.Value, error) {
	return id.String(), nil
}


gorilla.go
// MIT License
//
// Copyright (c) 2021 Iván Szkiba
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package nbid

import (
	"reflect"
)

// Convert is a custom converter function for github.com/gorilla/schema decoder.
func Convert(value string) reflect.Value {
	id, err := Parse(value)
	if err != nil {
		return invalidValue
	}

	return reflect.ValueOf(id)
}


cmd/nbid/main.go
// MIT License
//
// Copyright (c) 2021 Iván Szkiba
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package main

import (
	"flag"
	"fmt"
	"os"
	"runtime"

	"github.com/szkiba/nbid"
)

var version = "dev"

type options struct {
	version bool
	input   string
}

const usage = `usage: %s [name]

Generate NBID for name, or random NBID if name is missing.

Example: %s "The quick brown fox jumps over the lazy dog"
Output: QUKFNCO7QU098QEAJAUB021E9S

`

func getopt(args []string) *options {
	flags := flag.NewFlagSet(args[0], flag.ExitOnError)

	flags.Usage = func() {
		fmt.Fprintf(flags.Output(), usage, flags.Name(), flags.Name())
		flags.PrintDefaults()
	}

	o := options{}

	ver := flags.Bool("v", false, "prints version")

	_ = flags.Parse(args[1:])

	o.version = *ver
	o.input = flags.Arg(0)

	return &o
}

func getid(s string) string {
	if s == "" {
		return nbid.Random().String()
	}

	return nbid.New([]byte(s)).String()
}

func getver() string {
	return fmt.Sprintf("nbid/%s %s/%s", version, runtime.GOOS, runtime.GOARCH)
}

func main() {
	o := getopt(os.Args)

	if o.version {
		fmt.Fprintln(os.Stderr, getver())
	} else {
		fmt.Fprintln(os.Stdout, getid(o.input))
	}
}


id.go
// MIT License
//
// Copyright (c) 2021 Iván Szkiba
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.

package nbid

import (
	"bytes"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base32"
	"errors"
	"hash"
)

var (
	// ErrInvalidID is returned when trying to unmarshal an invalid NBID.
	ErrInvalidID = errors.New("nbid: invalid NBID")

	// Nil is an empty NBID (all zero).
	Nil NBID
)

const (
	encodedLen = 26 // string encoded len
	rawLen     = 16 // binary raw len
)

// NBID represents a Name Based ID.
// The binary representation of the NBID is a 16 byte byte array.
// The string representation is using base32 hex (w/o padding).
type NBID [rawLen]byte

// NewHash returns a new NBID derived from the hash of data generated by h.
// The hash should be at least 16 byte in length.
// The first 16 bytes of the hash are used to form the ID.
func NewHash(h hash.Hash, data []byte) NBID {
	h.Reset()
	h.Write(data) //nolint:errcheck
	s := h.Sum(nil)

	var id NBID

	copy(id[:], s)

	return id
}

// New returns a new NBID derived from the SHA256 hash of data.
// The first 16 bytes of the SHA256 hash are used to form the ID.
// It is the same as calling:
//
//  NewHash(sha256.New(), data)
func New(data []byte) NBID {
	return NewHash(sha256.New(), data)
}

// Random returns a new random generated NBID.
// The strength of the IDs is based on the strength of the crypto/rand
// package.
func Random() NBID {
	var id NBID

	rand.Read(id[:]) //nolint:errcheck

	return id
}

// Parse decodes s into an NBID or returns an error.
// The string representation is using base32 hex (w/o padding).
// Returns an error if the s does not have a length of 16.
func Parse(s string) (NBID, error) {
	i := &NBID{}
	err := i.UnmarshalText([]byte(s))

	return *i, err
}

// String returns the string form of NBID, a base32 hex (w/o padding).
// RFC 4648 / 7. Base 32 Encoding with Extended Hex Alphabet
// https://tools.ietf.org/html/rfc4648#section-7
func (id NBID) String() string {
	return base32.HexEncoding.WithPadding(base32.NoPadding).EncodeToString(id[:])
}

// MustParse decodes s into an NBID or panics if the string cannot be parsed.
func MustParse(s string) NBID {
	id, err := Parse(s)
	if err != nil {
		panic(err)
	}

	return id
}

// MarshalText implements encoding/text TextMarshaler interface.
func (id NBID) MarshalText() (text []byte, err error) {
	return []byte(id.String()), nil
}

// MarshalBinary implements encoding.BinaryMarshaler.
func (id NBID) MarshalBinary() ([]byte, error) {
	return id[:], nil
}

// MarshalJSON implements encoding/json Marshaler interface.
func (id NBID) MarshalJSON() ([]byte, error) {
	if id.IsNil() {
		return []byte("null"), nil
	}

	text, err := id.MarshalText()

	return []byte(`"` + string(text) + `"`), err
}

// UnmarshalText implements encoding/text TextUnmarshaler interface.
func (id *NBID) UnmarshalText(text []byte) error {
	if len(text) != encodedLen {
		return ErrInvalidID
	}

	_, err := base32.HexEncoding.WithPadding(base32.NoPadding).Decode(id[:], text)
	if err != nil {
		return ErrInvalidID
	}

	return nil
}

// UnmarshalBinary implements encoding.BinaryUnmarshaler.
func (id *NBID) UnmarshalBinary(data []byte) error {
	if len(data) != rawLen {
		return ErrInvalidID
	}

	copy(id[:], data)

	return nil
}

// UnmarshalJSON implements encoding/json Unmarshaler interface.
func (id *NBID) UnmarshalJSON(b []byte) error {
	s := string(b)
	if s == "null" {
		*id = Nil

		return nil
	}

	return id.UnmarshalText(b[1 : len(b)-1])
}

// IsNil returns true if this is a "nil" NBID.
func (id NBID) IsNil() bool {
	return id.Equal(Nil)
}

// Bytes returns the byte array representation of NBID.
func (id NBID) Bytes() []byte {
	return id[:]
}

// FromBytes creates a new NBID from a byte slice.
// Returns an error if the slice does not have a length of 16.
// The bytes are copied from the slice.
func FromBytes(b []byte) (NBID, error) {
	var id NBID

	if len(b) != rawLen {
		return id, ErrInvalidID
	}

	copy(id[:], b)

	return id, nil
}

// Compare returns an integer comparing two NBIDs. It behaves just like `bytes.Compare`.
func (id NBID) Compare(other NBID) int {
	return bytes.Compare(id[:], other[:])
}

// Equal returns true if two NBID is equal. It behaves just like `bytes.Equal`.
func (id NBID) Equal(other NBID) bool {
	return bytes.Equal(id[:], other[:])
}
